### ISSUE TITLE
每个epoch的mask都一样

### ISSUE THREAD

**Issue by yeyang16** [pengzhiliang/MAE-pytorch#28]
您好，在使用您的代码过程中发现每个epoch的mask都是一样的，目前替换成了下面的代码：
mask = torch.randperm(self.num_patches) < self.num_mask
mask = mask.long()

**Comment by PhoneSix** [comment]
您好，请问您是怎么确认每个epoch的mask都是一样的呢？我们这边用如下代码测试了一下，mask是不一样的。
```
g = RandomMaskingGenerator(224, 0.75)  
mask1 = g()  
mask2 = g()  
print(sum(abs(mask1-mask2)))  
```
输出不是0.

**Comment by pengzhiliang** [comment]
Please feel free to reopen this issue with more info if you are still stuck in this problem.
Thank you！

**Comment by wgting96** [comment]
I guess this bug may be related to the numpy random seed issue mentioned in https://github.com/pytorch/pytorch/issues/5059 .
The torch version >= 1.9 should be fine. Please correct me if I am wrong.

**Comment by PhoneSix** [comment]
I test it with pytorch 1.5.1, 1.7.1 and 1.10.0. All work well. @wgting96

**Comment by wgting96** [comment]
A minimal code snippet to reproduce:

```python
from torch.utils.data import Dataset, DataLoader
import numpy as np

class RandomMaskingGenerator:
    def __init__(self, input_size, mask_ratio):
        if not isinstance(input_size, tuple):
            input_size = (input_size,) * 2

        self.height, self.width = input_size

        self.num_patches = self.height * self.width
        self.num_mask = int(mask_ratio * self.num_patches)

    def __call__(self):
        mask = np.hstack([
            np.zeros(self.num_patches - self.num_mask),
            np.ones(self.num_mask),
        ])
        np.random.shuffle(mask)

        return mask # [196]

class DummyTest(Dataset):

    def __init__(self):
        self.masked_position_generator = RandomMaskingGenerator(3, 0.5)

    def __len__(self):
        return 4

    def __getitem__(self, item):
        return self.masked_position_generator()

if __name__ == '__main__':

    dataloader = DataLoader(DummyTest(), batch_size=1, num_workers=1)
    for epoch in range(4):
        print(f"==> Epoch {epoch}")
        for it, data in enumerate(dataloader):
            print(it, data)

``` 

Output:
```
python dummy_test.py
==> Epoch 0
0 tensor([[0., 1., 0., 0., 0., 1., 1., 1., 0.]], dtype=torch.float64)
1 tensor([[0., 1., 0., 0., 1., 0., 0., 1., 1.]], dtype=torch.float64)
2 tensor([[0., 0., 1., 0., 0., 0., 1., 1., 1.]], dtype=torch.float64)
3 tensor([[0., 0., 0., 1., 0., 1., 0., 1., 1.]], dtype=torch.float64)
==> Epoch 1
0 tensor([[0., 1., 0., 0., 0., 1., 1., 1., 0.]], dtype=torch.float64)
1 tensor([[0., 1., 0., 0., 1., 0., 0., 1., 1.]], dtype=torch.float64)
2 tensor([[0., 0., 1., 0., 0., 0., 1., 1., 1.]], dtype=torch.float64)
3 tensor([[0., 0., 0., 1., 0., 1., 0., 1., 1.]], dtype=torch.float64)
==> Epoch 2
0 tensor([[0., 1., 0., 0., 0., 1., 1., 1., 0.]], dtype=torch.float64)
1 tensor([[0., 1., 0., 0., 1., 0., 0., 1., 1.]], dtype=torch.float64)
2 tensor([[0., 0., 1., 0., 0., 0., 1., 1., 1.]], dtype=torch.float64)
3 tensor([[0., 0., 0., 1., 0., 1., 0., 1., 1.]], dtype=torch.float64)
==> Epoch 3
0 tensor([[0., 1., 0., 0., 0., 1., 1., 1., 0.]], dtype=torch.float64)
1 tensor([[0., 1., 0., 0., 1., 0., 0., 1., 1.]], dtype=torch.float64)
2 tensor([[0., 0., 1., 0., 0., 0., 1., 1., 1.]], dtype=torch.float64)
3 tensor([[0., 0., 0., 1., 0., 1., 0., 1., 1.]], dtype=torch.float64)

python -c "import torch; print(torch.__version__)"
1.8.0
```

**Comment by pengzhiliang** [comment]
Thanks for your report and suggestion! 

We will test it and fix it later!

> mask = torch.randperm(self.num_patches) < self.num_mask
mask = mask.long()

Thank u again!

**Comment by pengzhiliang** [comment]
As far as I know, in the training process, the image order in each epoch is different.
Therefore, even if the mask is the same in different epoch, but the corresponding images are different, the results are still different.

**Comment by yeyang16** [comment]
Yeah, but different masks will be better. Thank all those who pay attention to this issue

**Comment by ayumiymk** [comment]
> Yeah, but different masks will be better. Thank all those who pay attention to this issue

Hi, do you find any solution?

**Comment by ayumiymk** [comment]
Actually, I find when the num_workers > 0, the generated mask is also the same every num_workers times. Then I find a solution [here](https://pytorch.org/docs/stable/notes/randomness.html#dataloader).

**Comment by pengzhiliang** [comment]
@ayumiymk I tested the solution:

```python
def seed_worker(worker_id):
    worker_seed = torch.initial_seed() % 2**32
    np.random.seed(worker_seed)
    random.seed(worker_seed)

dataloader = DataLoader(DummyTest(), batch_size=1, num_workers=1, worker_init_fn=seed_worker)
for epoch in range(4):
    print(f"==> Epoch {epoch}")
    for it, data in enumerate(dataloader):
        print(it, data)
```
and got the output:
```
==> Epoch 0
0 tensor([[0., 0., 1., 0., 1., 1., 0., 1., 0.]], dtype=torch.float64)
1 tensor([[1., 0., 0., 1., 0., 0., 1., 0., 1.]], dtype=torch.float64)
2 tensor([[0., 1., 0., 0., 0., 1., 1., 1., 0.]], dtype=torch.float64)
3 tensor([[0., 0., 1., 0., 0., 1., 1., 1., 0.]], dtype=torch.float64)
==> Epoch 1
0 tensor([[1., 1., 1., 0., 0., 0., 1., 0., 0.]], dtype=torch.float64)
1 tensor([[1., 0., 0., 1., 1., 0., 1., 0., 0.]], dtype=torch.float64)
2 tensor([[0., 1., 1., 0., 0., 0., 1., 0., 1.]], dtype=torch.float64)
3 tensor([[0., 1., 0., 0., 0., 0., 1., 1., 1.]], dtype=torch.float64)
==> Epoch 2
0 tensor([[1., 0., 0., 0., 0., 1., 1., 0., 1.]], dtype=torch.float64)
1 tensor([[0., 0., 1., 0., 0., 1., 0., 1., 1.]], dtype=torch.float64)
2 tensor([[0., 0., 0., 1., 1., 1., 0., 1., 0.]], dtype=torch.float64)
3 tensor([[0., 1., 0., 1., 1., 0., 1., 0., 0.]], dtype=torch.float64)
==> Epoch 3
0 tensor([[1., 0., 0., 1., 0., 1., 0., 1., 0.]], dtype=torch.float64)
1 tensor([[1., 1., 0., 0., 1., 0., 1., 0., 0.]], dtype=torch.float64)
2 tensor([[0., 0., 0., 1., 1., 1., 0., 1., 0.]], dtype=torch.float64)
3 tensor([[1., 0., 0., 0., 1., 1., 0., 0., 1.]], dtype=torch.float64)
```

It seems different. 
Thank you very much!

**Comment by pengzhiliang** [comment]
It's been fixed. Thank you!

https://github.com/pengzhiliang/MAE-pytorch/commit/e4e21b7d04a012046fb920a9b758bab415739d20

**Comment by PistonY** [comment]
> 您好，在使用您的代码过程中发现每个epoch的mask都是一样的，目前替换成了下面的代码： mask = torch.randperm(self.num_patches) < self.num_mask mask = mask.long()

这个实现实在是太优雅了
